README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

@muschellij2 badges: [![Travis build
status](https://travis-ci.com/muschellij2/ActivityIndex.svg?branch=master)](https://travis-ci.com/muschellij2/ActivityIndex)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/muschellij2/ActivityIndex?branch=master&svg=true)](https://ci.appveyor.com/project/muschellij2/ActivityIndex)

# ActivityIndex

ActivityIndex is an R package which provides functions to read and
process raw accelerometry data.

## Overview

ActivityIndex provides R functions to read raw accelerometry data
collected by accelerometers. Specifically, it can directly handle
**.csv** data files generated by accelerometer model GT3X+ and software
ActiLife by ActiGraph. The package also provides functions to calculate
summarizing metrics such as Activity Index (AI), using the raw data.

The AI is a way of summarizing densely sampled accelerometry data into
given epochs (such as every 1 second or every 15 seconds, etc).
Essentially, AI describes the variability of the raw acceleration
signals, after normalizing it using systematic noise of the signal. AI
is an evolution of the original metric, **Activity Intensity**, proposed
in the paper [Normalization and extraction of interpretable metrics from
raw accelerometry data](https://doi.org/10.1093/biostatistics/kxt029) by
J. Bai et al (2014). The AI addresses some limitation of the original
Activity Intensity, and has favorable properties. More details on these
properties and a direct comparison of AI versus other metrics such as
Activity Count could be found in the paper [An Activity Index for Raw
Accelerometry Data and Its Comparison with Other Activity
Metrics](https://doi.org/10.1371/journal.pone.0160644).

## Package Installation

ActivityIndex software can be installed via GitHub. Users should have
`R` installed on their computer before installing ActivityIndex.

### Install devtools package

To install ActivityIndex package via GitHub, the user must have
installed [devtools](https://cran.r-project.org/package=devtools), which
could be completed by using the following R command

``` r
install.packages("devtools")
```

### Install ActivityIndex package

The following R command can be used to install the latest version of
ActivityIndex via GitHub:

``` r
devtools::install_github("javybai/ActivityIndex")
```

## ActivityIndex User Manual

The ActivityIndex package includes a vignette to demonstrate a typical
work flow of computing AI. The vignette can either be accessed by R
command

``` r
browseVignettes(package="ActivityIndex")
```

and interactively browsing or going to
<https://javybai.github.io/ActivityIndex/articles/ActivityIndexIntro.html>.

## Further Questions

Please contact the author and maintainer Jiawei Bai (jbai \[at\] jhsph
\[dot\] edu) or open an issue at GitHub.


NEWS.md
# ActivityIndex 0.3.7

* S3 dispatch for computeActivityIndex and GT3XPlus object now done.
* Fixed typos
* Fixed noLD issues in the examples

# ActivityIndex 0.3.6

* Example for `computeActivityIndex` added, taken from vignette.
* `print` methods for S3 `ActivityIndex` object created.

# ActivityIndex 0.3.5

* Vignette now in Markdown.
* Site is built using `pkgdown::build_site`.
* CRAN release version.

# ActivityIndex 0.3.4

* Adding tests with `testthat`.
* Added conversions to `data.table` when necessary.  
* Fixing docs with `\code` accidentally in there.

# ActivityIndex 0.3.3

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
library(testthat)
library(ActivityIndex)

test_check("ActivityIndex")


tests/testthat/test-Sigma0.R
testthat::context("Simple test for Sigma0")

testthat::test_that("Make sure Sigma0 giving same answer", {
  filename = system.file("extdata","sample_GT3X+.csv.gz",package="ActivityIndex")
  res = ReadGT3XPlus(filename)
  hertz = res$Hertz
  x = res$Raw[ 1:1000, c("Time", "X", "Y", "Z")]
  res = Sigma0(x, hertz = hertz)
  testthat::expect_equal(res, c(SD = 0.184321637135534))

  # tests for data.frame
  res_df = Sigma0(as.data.frame(x), hertz = hertz)
  testthat::expect_equal(res, res_df)
})


tests/testthat/test-computeAI.R
testthat::context("Output doesn't change for the Sample Data")

testthat::test_that("computeAI same answer regardless of input", {

  sampleGT3XPlus = ReadGT3XPlus(system.file("extdata","sample_GT3X+.csv.gz",
                                            package="ActivityIndex"))
  sampleTable = ReadTable(system.file("extdata", "sample_table.csv.gz",
                                      package="ActivityIndex"))

  testthat::expect_named(sampleGT3XPlus, c("SN", "StartTime", "StartDate",
                                           "Epoch", "DownloadTime",
                                           "DownloadDate",
                                           "Hertz", "Raw"))

  testthat::expect_named(sampleGT3XPlus$Raw, c("Date", "Time", "X", "Y", "Z"))

  s0 = sampleTable[1004700:1005600,]
  AI_sampleTable_x = computeActivityIndex(
    sampleTable,
    x_sigma0=s0, epoch=1, hertz=30)
  AI_sampleGT3XPlus_x = computeActivityIndex(
    sampleGT3XPlus, x_sigma0=s0,
    epoch=1,
    hertz=30)

  testthat::expect_silent(
    computeActivityIndex(sampleGT3XPlus,
                         epoch = 1L,
                         hertz = 30L,
                         sigma0 = 0)
  )

  testthat::expect_equal(AI_sampleTable_x$AI,
                         AI_sampleGT3XPlus_x$AI)
  testthat::expect_error(
    computeActivityIndex(sampleGT3XPlus, x_sigma0=s0,
                         epoch = 1,
                         hertz = 100)
  )
  testthat::expect_error(
    computeActivityIndex(sampleGT3XPlus, x_sigma0=s0,
                         epoch = 0.3,
                         hertz = 100),
    regexp = "must not be less"
  )
  testthat::expect_error(
    computeActivityIndex(sampleGT3XPlus, x_sigma0=s0,
                         epoch = 1.5,
                         hertz = 100),
    regexp = "must be an integer"
  )
  x = AI_sampleGT3XPlus_x$AI
  stats = c(mean(x), sd(x), median(x),
            mean(x[x > 0]), median(x[x > 0]))
  testthat::expect_equal(stats,
                         c(22.7449457740193, 67.1315848298163,
                           0.961727793460805, 36.1350372579237,
                           4.33613021194137))
})


R/Sigma0.r
#' @title \eqn{\bar{\sigma}_i} computing using raw accelerometry data
#' @description \code{Sigma0} computes \eqn{\bar{\sigma}_i}, which is needed for
#'  the Activity Index computing in \code{\link{computeActivityIndex}}
#' @param x A 4-column data frame containing the raw accelerometry
#'  data when the device is not worn. The 1st column has the record/index
#'  number. The 2nd to 4th columns contain the tri-axial raw acceleration. The
#'  data will be used to calculate \eqn{\bar{\sigma}_i}.
#' @param hertz The sample rate of the data.
#' @return \eqn{\bar{\sigma}_i}, a numeric vector of length one.
#' @export
#' @examples
#' filename = system.file("extdata","sample_GT3X+.csv.gz",package="ActivityIndex")
#' res = ReadGT3XPlus(filename)
#' hertz = res$Hertz
#' x = res$Raw[ 1:1000, c("Time", "X", "Y", "Z")]
#' res = Sigma0(x, hertz = hertz)
#' testthat::expect_equal(res, c(SD = 0.184321637135534))
Sigma0 = function(x, hertz = 30)
{
  # stopifnot(is.data.table(x))
  x = as.data.table(x)
  X = Y = Z = NULL
  rm(list = c("X", "Y", "Z"))
  n = nrow(x) %/% hertz*hertz
  result = mean(x[1:n,
                sqrt((rowVars(matrix(X, ncol = hertz, byrow=TRUE))+
                        rowVars(matrix(Y, ncol = hertz, byrow=TRUE))+
                        rowVars(matrix(Z, ncol = hertz, byrow=TRUE)))/3)]
                )
  names(result) = c("SD")
  return(result)
}


R/ReadTable.r
#' @title Read the raw tri-axial accelerometry data csv file.
#'
#' @description
#' \code{ReadTable} reads the raw tri-axial accelerometry data in csv files.
#'
#' @details
#' The function reads csv files containing only three columns: acceleration time series
#' of x-, y- and z-axis.
#'
#' @param filename
#' The name of the csv file.
#'
#' @return The \code{ReadTable} returns a data frame with 4 columns: Index, X, Y and Z.
#' Index is the column for the indices of acceleration. X, Y and Z are for the acceleration
#' time series in each direction.
#' @export
#' @importFrom utils read.csv
#' @examples
#' filename = system.file("extdata","sample_table.csv.gz",package="ActivityIndex")
#' res = ReadTable(filename)
#'
ReadTable = function(filename) {
  if (ncol(read.csv(file = filename,
                    stringsAsFactors = FALSE,
                    header = FALSE,
                    nrows = 1)) != 3)
  {
    stop(paste0(filename, " is not an appropriate 3-column data file!"))
  }
  result = fread(
    filename,
    sep = ",",
    stringsAsFactors = FALSE,
    colClasses = rep("numeric", 3),
    header = FALSE,
    showProgress = FALSE
  )
  result = cbind(1:nrow(result), result)
  colnames(result) = c("Index", "X", "Y", "Z")
  return(result)
}


R/ReadGT3XPlus.r
#' @title Read the raw tri-axial accelerometry data csv file by ActiGraph GT3X+.
#'
#' @description
#' \code{ReadGT3XPlus} reads the accelerometry data collected by ActiGraph GT3XPlus
#' in csv files generated by ActiLife software. It automatically parses the header
#' of the csv file to acquire the setting of the device.
#'
#' @details
#' The function is tested on the csv files generated by ActiLife6, which have exactly
#' 10 lines of headers, containing information about the device name, the starting
#' and ending date/time of data collection, the sample rate, and the downloading date/time,
#' etc. The 11th may further be omitted, if it is the header for the tri-axial acceleration
#' time series. The function only reads the first 3 columns from then, if more are present.
#'
#' @param filename
#' The name of the csv file.
#'
#' @param epoch
#' The epoch length (in second) of the Activity Index. Must
#' be a positive integer. The default is \code{1}.
#'
#' @return The \code{ReadGT3XPlus} returns an object of \link{class} "\code{GT3XPlus}".
#' This class of object is supported by functions \code{\link{computeActivityIndex}}.
#' A object of class "\code{GT3XPlus}" is a list containing at least the following components:
#'
#'  \code{SN}: Serial Number of the accelerometer
#'
#'  \code{StartTime}: Start time of the data collection
#'
#'  \code{StartDate}: Start date of the data collection
#'
#'  \code{Epoch}: The Epoch time of each observation. If sample rate \code{Hertz}>1, then \code{Epoch}=00:00:00
#'
#'  \code{DownloadTime}: Download time of the data
#'
#'  \code{DownloadDate}: Download date the data
#'
#'  \code{Hertz}: Sampling Rate
#'
#'  \code{Raw}: a data frame with 5 columns containing the date, time and acceleration in X, Y and Z axes
#' for each observation.
#'
#' @export
#'
#' @import data.table
#' @import matrixStats
#'
#' @importFrom R.utils gunzip
#' @examples
#' filename = system.file("extdata","sample_GT3X+.csv.gz",package="ActivityIndex")
#' res = ReadGT3XPlus(filename)
#'
ReadGT3XPlus = function(filename, epoch=1) {
  stopifnot(length(filename) == 1)
  if (grepl("[.]gz$", filename)) {
    filename = R.utils::gunzip(filename,
                               temporary = TRUE, remove = FALSE,
                               overwrite = TRUE)
  }
  result=list(SN="",StartTime="",StartDate="",Epoch="",DownloadTime="",DownloadDate="",Hertz="",Raw="")
  ### Header ###
  result_Head=readLines(filename,10)
  result$Epoch=regmatches(result_Head[5],regexpr("(?<=Epoch\\sPeriod\\s\\(hh\\:mm\\:ss\\)\\s)(.+)(?=\\Z)",result_Head[5],perl=TRUE))
  result$Hertz=ifelse(length(as.numeric(regmatches(result_Head[1],regexpr("(?<=\\s)(\\d+)(?=\\sHz)",result_Head[1],perl=TRUE))))>0,
                      as.numeric(regmatches(result_Head[1],regexpr("(?<=\\s)(\\d+)(?=\\sHz)",result_Head[1],perl=TRUE))),
                      1)
  result$SN=regmatches(result_Head[2],regexpr("(?<=Serial Number:\\s)(.+)(?=\\Z)",result_Head[2],perl=TRUE))
  result$StartTime=regmatches(result_Head[3],regexpr("(?<=Start\\sTime\\s)(.+)(?=\\Z)",result_Head[3],perl=TRUE))
  result$StartDate=regmatches(result_Head[4],regexpr("(?<=Start\\sDate\\s)(.+)(?=\\Z)",result_Head[4],perl=TRUE))
  result$StartDate=as.character(as.Date(result$StartDate,format="%m/%d/%Y"))
  result$DownloadTime=regmatches(result_Head[6],regexpr("(?<=Download\\sTime\\s)(.+)(?=\\Z)",result_Head[6],perl=TRUE))
  result$DownloadDate=regmatches(result_Head[7],regexpr("(?<=Download\\sDate\\s)(.+)(?=\\Z)",result_Head[7],perl=TRUE))
  result$DownloadDate=as.character(as.Date(result$DownloadDate,format="%m/%d/%Y"))
  ### Data ###
  result$Raw=read.csv(file=filename,skip=10,stringsAsFactors=FALSE,header=FALSE,nrows=1)
  if (is.character(result$Raw[,1])==TRUE)
  {
    row1=read.csv(file=filename,skip=11,stringsAsFactors=FALSE,header=FALSE,nrows=1)
    result$Raw=fread(filename,skip=11,sep=",",stringsAsFactors=FALSE,
                     colClasses=rep("numeric",ncol(row1)),header=FALSE,fill=TRUE,
                     showProgress=FALSE)
  } else
  {
    row1=read.csv(file=filename,skip=10,stringsAsFactors=FALSE,header=FALSE,nrows=1)
    result$Raw=fread(filename,skip=10,sep=",",stringsAsFactors=FALSE,
                     colClasses=rep("numeric",ncol(row1)),header=FALSE,fill=TRUE,
                     showProgress=FALSE)
  }
  TimeScale = ActivityIndex::TimeScale
  # Remove incomplete row
  if (any(is.na(result$Raw[nrow(result$Raw),]))) result$Raw=result$Raw[-nrow(result$Raw),]
  # Time Stamp #
  if (result$Epoch=="00:00:00")
  {
    Time_Temp_idx=which(TimeScale==result$StartTime):(which(TimeScale==result$StartTime)-1+nrow(result$Raw)%/%result$Hertz+1)
    Time_Temp_idx=Time_Temp_idx%%length(TimeScale)
    Time_Temp_idx=rep(Time_Temp_idx,each=result$Hertz)
    Time_Temp_idx=Time_Temp_idx[1:nrow(result$Raw)]
    Time_Temp_idx[which(Time_Temp_idx==0)]=length(TimeScale)
  } else
  {
    Time_Temp_idx=which(TimeScale==result$StartTime):(which(TimeScale==result$StartTime)-1+nrow(result$Raw)*epoch+1)
    Time_Temp_idx=Time_Temp_idx%%length(TimeScale)
    Time_Temp_idx=Time_Temp_idx[which((1:length(Time_Temp_idx))%%epoch==1)]
    Time_Temp_idx=Time_Temp_idx[1:nrow(result$Raw)]
    Time_Temp_idx[which(Time_Temp_idx==0)]=length(TimeScale)
  }
  # Combine #
  result$Raw=cbind(rep(result$StartDate,nrow(result$Raw)),TimeScale[Time_Temp_idx],result$Raw)
  if (ncol(result$Raw)>5)
  {
    colnames(result$Raw)=c("Date","Time","X","Y","Z",paste0("V",1:(ncol(result$Raw)-5)))
  } else
  {
    colnames(result$Raw)=c("Date","Time","X","Y","Z")
  }

  # Date Stamp
  # Change the Date if sample reaches midnight
  if (length(which(result$Raw$Time=="00:00:00"))>0)
  {
    date_idx_start=which(result$Raw$Time=="00:00:00")[(1:(length(which(result$Raw$Time=="00:00:00"))%/%result$Hertz)-1)*result$Hertz+1]
    date_idx_end=c(date_idx_start[-1]-1,nrow(result$Raw))
    if (date_idx_start[1]==1)
    {
      date_follow=as.character(as.Date(result$StartDate)+1:length(date_idx_start)-1)
    } else
    {
      date_follow=as.character(as.Date(result$StartDate)+1:length(date_idx_start))
    }
    for (i in 1:length(date_idx_start))
    {
      result$Raw$Date[date_idx_start[i]:date_idx_end[i]]=rep(date_follow[i],length(date_idx_start[i]:date_idx_end[i]))
    }
  }
  #
  class(result)="GT3XPlus"
  return(result)
}


R/computeActivityIndex.r
#' @title Compute Activity Index
#' @description \code{computeActivityIndex} computes the Activity Index using raw
#' accelerometry data, based on user specified parameters such as sample rate
#' and epoch length.
#'
#' @param x An object containing raw accelerometry data, which could
#' either be a 4-column data frame or "\code{GT3XPlus}" object. See "Details".
#' @param x_sigma0 A 4-column data frame containing the raw accelerometry
#'  data when the device is not worn. The 1st column has the record/index
#'  number. The 2nd to 4th columns contain the tri-axial raw acceleration. The
#'  data will be used to calculate \eqn{\bar{\sigma}_i}.
#' @param sigma0 Specify \eqn{\bar{\sigma}_i} directly. At least one of
#' \code{x_sigma0} and \code{sigma0} must be specified. If both existed,
#' \code{sigma0} will be used.
#' @param epoch The epoch length (in second) of the Activity Index. Must
#' be a positive integer.
#' @param hertz The sample rate of the data.
#' @return A data frame with two columns. The first column has the "record
#' number" associated with each entry of Activity Index, while the second column
#'  has the actual value of Activity Index.
#' @details \code{x} could be either of the following two types of objects:
#' \enumerate{
#'    \item A 4-column data frame containing the tri-axial raw accelerometry
#'    data in the 2nd to 4th column, and the associated record number (could be
#'    time) in the 1st column. \code{\link{ReadTable}} can be used to generate
#'    such data frame.
#'    \item An "\code{GT3XPlus}" object given by function
#'    \code{\link{ReadGT3XPlus}}.
#'    }
#' @examples
#' library(graphics)
#' fname = system.file("extdata", "sample_table.csv.gz",
#' package = "ActivityIndex")
#' sampleTable = ReadTable(fname)
#' AI_sampleTable_x = computeActivityIndex(
#'   sampleTable,
#'   x_sigma0 = sampleTable[1004700:1005600, ],
#'   epoch = 1,
#'   hertz = 30)
#' AI_sampleTable_x
#' plot(AI ~ RecordNo, data = AI_sampleTable_x, type = "l")
#' @export
computeActivityIndex=function(x,x_sigma0=NULL,sigma0=NULL,epoch=1,hertz)
{
  if (epoch<1) stop("epoch must not be less than 1!")
  if (abs(round(epoch)-epoch)>0) stop("epoch must be an integer!")
  UseMethod("computeActivityIndex",x)
}


#' @rdname computeActivityIndex
#' @export
computeActivityIndex.default=function(x,x_sigma0=NULL,sigma0=NULL,epoch=1,hertz)
{
  Index = NULL
  rm("Index")
  if (is.null(x_sigma0)&&is.null(sigma0)) stop("Either x_sigma0 or sigma0 needs to be specified!")
  if (is.null(sigma0))
  {
    sigma0=Sigma0(x_sigma0,hertz)
  }
  x = as.data.table(x)
  n=nrow(x)%/%hertz*hertz
  result=array(0,c(n%/%hertz,2))
  if (sigma0!=0)
  {
    result[,2]=x[1:n,(rowSds(matrix(x$X,ncol=hertz,byrow=TRUE))^2-sigma0^2)/sigma0^2+
                      (rowSds(matrix(x$Y,ncol=hertz,byrow=TRUE))^2-sigma0^2)/sigma0^2+
                      (rowSds(matrix(x$Z,ncol=hertz,byrow=TRUE))^2-sigma0^2)/sigma0^2]
  } else
  {
    result[,2]=x[1:n,rowSds(matrix(x$X,ncol=hertz,byrow=TRUE))^2+
                        rowSds(matrix(x$Y,ncol=hertz,byrow=TRUE))^2+
                        rowSds(matrix(x$Z,ncol=hertz,byrow=TRUE))^2]
  }
  result[which(result[,2]<0),2]=0
  result[,2]=sqrt(result[,2]/3)
  result=as.data.frame(result,stringsAsFactors=FALSE)
  # is Index a column name or some other thing?
  result[,1]=x[(1:(n%/%hertz)-1)*hertz+1,Index] # fetch "Index"
  if (epoch>1)
  {
    L_AI=length(result[,2])
    result0=as.data.frame(array(0,c(L_AI%/%epoch,2)),stringsAsFactors=FALSE)
    result0[,2]=as.numeric(rowSums(matrix(result[1:(L_AI-L_AI%%epoch),2],ncol=epoch,byrow=TRUE)))
    result0[,1]=result[(1:(L_AI%/%epoch)-1)*epoch+1,1]
    result=result0
  }
  colnames(result)=c("RecordNo","AI")
  class(result) = c("ActivityIndex", class(result))
  return(result)
}

#' @rdname computeActivityIndex
#' @export
computeActivityIndex.GT3XPlus=function(x,x_sigma0=NULL,sigma0=NULL,epoch=1,hertz)
{
  if (x$Hertz != hertz) {
    stop("hertz must be equal to the sample rate of GT3XPlus!")
  }
  x = x$Raw
  cn = colnames(x)
  cn[ cn == "Time"] = "Index"
  colnames(x) = cn

  result = computeActivityIndex.default(x,
                               x_sigma0 = x_sigma0,
                               sigma0 = sigma0,
                               epoch = epoch,
                               hertz = hertz)
  return(result)
}
